futures-util/src/await/select.rs

```diff
@@ -0,0 +1,49 @@
+//! The `select` macro.
+
+#[macro_export]
+macro_rules! select {
+    ($(
+        $name:ident => $body:expr,
+    )*) => { {
+        $(
+            $crate::await::assert_unpin(&$name);
+            let mut $name = $crate::future::maybe_done(&mut $name);
+            let mut $name = ::core::mem::PinMut::new(&mut $name);
+        )*
+        loop {
+            $(
+                if let ::core::task::Poll::Ready(()) = poll!($name.reborrow()) {
+                    break;
+                }
+            )*
+            pending!();
+        }
+        if false {
+            unreachable!()
+        }
+        $(
+            else if let Some($name) = $name.take_output() {
+                $body
+            }
+        )*
+        else {
+            unreachable!()
```

Comment:
Note: this is actually reachable by writing `select!{}`-- perhaps worth a special panic message?